python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatchPointsTo.qll

```diff
@@ -533,26 +767,51 @@ class ReturnKind extends TReturnKind {
 }
 
 /** A data flow node that represents a value returned by a callable. */
-class ReturnNode extends CfgNode {
-  Return ret;
+abstract class ReturnNode extends Node {
+  /** Gets the kind of this return node. */
+  ReturnKind getKind() { any() }
+}
 
+/** A data flow node that represents a value returned by a callable. */
+class ReturnSourceNode extends ReturnNode, CfgNode {
   // See `TaintTrackingImplementation::returnFlowStep`
-  ReturnNode() { node = ret.getValue().getAFlowNode() }
+  ReturnSourceNode() { node = any(Return ret).getValue().getAFlowNode() }
 
-  /** Gets the kind of this return node. */
-  ReturnKind getKind() { any() }
+  override ReturnKind getKind() { any() }
 }
 
-/** A data flow node that represents the output of a call. */
-class OutNode extends CfgNode {
-  OutNode() { node instanceof CallNode }
+/** A data-flow node that represents the output of a call. */
+abstract class OutNode extends Node {
+  /** Gets the underlying call, where this node is a corresponding output of kind `kind`. */
+  abstract DataFlowCall getCall(ReturnKind kind);
+}
+
+private module OutNodes {
+  /**
+   * A data-flow node that reads a value returned directly by a callable.
+   */
+  class ExprOutNode extends OutNode, ExprNode {
```

Comment:
Why not extend CallCfgNode? or at least CfgNode? -- again, I feel like it goes against our practice to drop all the way down to the expression, instead of staying in the control flow layer.

Reply:
`ExprNode` is a control flow node, but you are right that the charred is defined in terms of expressions. I have amended that.